keps/sig-node/4381-dra-structured-parameters/README.md

```diff
@@ -1475,225 +1382,400 @@ prevented. Deleting the entire cluster also leaves resources allocated outside
+    // all ordinary claims to the device with respect to access modes and
+    // any resource allocations. Ability to request this kind of access is
+    // controlled via ResourceQuota in the resource.k8s.io API.
+    //
+    // Default is false.
     //
-    // There can be at most 32 such reservations. This may get increased in
-    // the future, but not reduced.
     // +optional
-    ReservedFor []ResourceClaimConsumerReference
+    AdminAccess *bool
 }
+```
 
-// ReservedForMaxSize is the maximum number of entries in
-// claim.status.reservedFor.
-const ResourceClaimReservedForMaxSize = 32
 ```
+<<[UNRESOLVED @johnbelamaric ]>>
+
+There is nothing in the current revision of the KEP which depends on device classes
+being mandatory, therefore the `DeviceClassName` is optional. Before beta we need
+to decide whether we keep that or make it required. Changing after beta will be
+harder.
 
-##### ResourceClaimParameters
+<<[/UNRESOLVED]>>
+```
 
 ```go
-// ResourceClaimParameters defines resource requests for a ResourceClaim in an
-// in-tree format understood by Kubernetes.
-type ResourceClaimParameters struct {
-    metav1.TypeMeta
-    // Standard object metadata
-    metav1.ObjectMeta
 
-    // If this object was created from some other resource, then this links
-    // back to that resource. This field is used to find the in-tree representation
-    // of the claim parameters when the parameter reference of the claim refers
-    // to some unknown type.
-    GeneratedFrom *ResourceClaimParametersReference
+// Exactly one field must be set.
+type Selector struct {
+    // This CEL expression must evaluate to true if a
+    // device is suitable. This covers qualitative aspects of
+    // device selection.
+    //
+    // The language is as defined in
+    // https://kubernetes.io/docs/reference/using-api/cel/
+    // with several additions that are specific to device selectors.
+    //
+    // Attributes of a device are made available through a `device.attributes` map.
+    // The value of each entry varies, depending on the attribute
+    // that is being looked up.
+    //
+    // Unknown keys trigger a runtime error.
+    //
+    // The `device.driverName` string variable can be used to check for a specific
+    // driver explicitly in a filter that is meant to work for devices from
+    // different vendors. It is provided by Kubernetes and matches the
+    // `driverName` from the ResourceSlice which provides the device.
+    //
+    // The CEL expression is applied to *all* available devices from any driver.
+    // The expression has to check for existence of an attribute when it is not
+    // certain that it is provided because runtime errors are not automatically
+    // treated as "don't select device". Instead, device selection fails completely
+    // and reports the error.
+    //
+    // Some examples:
+    //    "memory.dra.example.com" in device.attributes && # Is the attribute available?
```

Comment:
You mean the comment?

Reply:
I think we are going to want to qualify them on the claim side, always. It's long but it will enable things like the composite device driver. If we end up one day with a cross-vendor `*.k8s.io/` model, maybe we can drop those in the claim side.

But here shouldn't it be `dra.example.com/memory` in any case?

Reply:
I thought we had landed on NOT qualifying them, which implied the drivername prefix?

Reply:
We don't require them to be qualified in the ResourceSlice where the driver name is the fallback domain prefix. But in the CEL expression we agreed to always do fully-qualified because the intended driver name cannot be inferred reliably.

> But here shouldn't it be dra.example.com/memory in any case?

Yes. Will fix.

Reply:
oh.  Yuck.  I guess I missed this incongruity.  Is there no way we can avoid this?  The CEL is going to be long enough as it is.

Reply:
Suppose we have a CEL expression `"something" in device.attributes && device.attributes["something"]` (expected type a boolean) and one device with driver name `dra.example.com`, `something`: false (= `dra.example.com/something`), `dra.k8s.io/something`: true.

Those `something` attributes have the same ID, but mean different things. Which one should be looked up by the CEL expression?

Suppose we have two different vendors which both use `something` for different things, potentially even with different types (boolean as expected, string for the other vendor). If we say "pick the vendor-qualified attribute", then the CEL above is potentially invalid because it doesn't return a boolean for the second vendor. This could be fixed by adding a `device.driverName` check, but it's not obvious that the check is missing. It'll work without that check on clusters with just that one driver and fail on clusters where the other driver is also installed.

Requiring fully-qualified names makes the CEL longer, but it avoids a lot of foot guns.

Reply:
I assume that the vast majority of claims WILL have vendor checks, so the only time you need a fully qualified name is when you are comparing across vendors, and if you are comparing across vendors you probably want all your references to be fully qualified.

This is safe:
    `dev.driver == "example.com" && dev.attrs["model"] == "foo" && devattrs["version"] > "1.2.3"` 

This is unsafe:
    `dev.attrs["model"] == "foo" && devattrs["version"] > "1.2.3"` 

At risk of re-adjudicating, perhaps this is a place and reason to consider the more "nested" CEL structure, where each prefix forms a new map?  If you combine that with the `bind` (I think that was the name, I'll be damned if I can find the spec Joe linked me to in the past) idea, you get something like:

`bind(attrs, dev.attrs["example.com"], attrs.model == "foo" && attrs.version > "1.2.3"` -- the names are short and concise, but are CLEARLY intended to be vendor specific.  If you don't like the field style, we can still do:

`bind(attrs, dev.attrs["example.com"], attrs["model"] == "foo" && attrs["version"] > "1.2.3"` 

This could also fit with CEL variable:

```
selectors:
- cel:
  variables:
  - name: "xattrs"
    expresion: dev.attrs["example.com"]
  expression: xattrs["model"] == "foo" && xattrs["version"] > "1.2.3"` 
```

Reply:
I think qualifications are needed (see https://github.com/kubernetes-sigs/wg-device-management/pull/27#issuecomment-2155167940 for example).

I like the map of maps. I think it gets us the best of both worlds. Is there any reason not to allow the variable construct in the API in 1.31?

Reply:
If we allow variables, would it be possible for the class to define a set of variables that can then used in the claim's expressions?

Reply:
> At risk of re-adjudicating, perhaps this is a place and reason to consider the more "nested" CEL structure, where each prefix forms a new map?

I'm fine with that. So basically `dev.attrs[<domain>][<id>]`, with `bind` or variables to reduce typing. Do we want to allow `dev.attrs[<domain>/id]`? It's unambiguous what the CEL is looking up (nested map or value) and it may help with simple cases (like `"dra.example.com/driverVersion" in dev.attrs`).

I don't have a strong opinion about field lookup vs. map for the leaf. Field lookup seems nicer, but we then have a mixture of styles :shrug:

> would it be possible for the class to define a set of variables that can then used in the claim's expressions?

Seems useful.

Reply:
field-style would mean both a mixture of styles, and going back to C-style identifiers. I if we do map-style lookups we should not do `attrs["<domain>/id"]` anymore. Just `attrs["domain"]["id"]`.

Reply:
> If we allow variables, would it be possible for the class to define a set of variables that can then used in the claim's expressions?

With required class (which we have now but I think Patrick was still uncertain about) that seems feasible. If class is not required, this won't work.

Reply:
> If we allow variables, would it be possible for the class to define a set of variables that can then used in the claim's expressions?

Maybe, but I don't know we would want that?  It seems like a recipe for problems that are hard to debug.  Once you publish a class with a variable, that can't be easily changed.   It becomes part of your class API.  I suppose it would have to be treated like lexically scoped vars.  E.g. class defines "foo" and class expression can use that variable.  claim expression can use it, but can also redefine it (at a smaller scope).  It seems like a lot of hassle - we should let the use-cases emerge. 

> I'm fine with that. So basically `dev.attrs[<domain>][<id>]`

It's the double-map that I find weird.  We don't allow the `id` to have slashes or dots (so no triple maps).  IIUC we cannot prevent field lookups, so `dev.attrs["example.com"]["attrName"]` and `dev.attrs["example.com"].attrName` would work but only `dev.attrs["example.com"]["attr-name"]` (hence the earlier "C identifier" format for the leaf).

I'm really shooting for "as simple as we can get away with", meaning less options and less noise.

* Option 1: What we have specced today
Every attr in CEL is `dev.attrs["example.com/attrName"]` or `dev.attrs["example.com/attr-name"]`

* Option 2: Split attr names on `/` and document only field lookups
Attr names are C identifiers.  Every attr in CEL is `dev.attrs["example.com"].attrName` (making "attr-name" invalid).  CEL variables can ecapsulate the `dev.attrs["example.com"]` part, yielding `xattrs.attrName`.

* Option 3: Split attr names on `/` and document only map lookups
Attr names are DNS labels (ish). Every attr in CEL is `dev.attrs["example.com"]["attrName"]` or `dev.attrs["example.com"]["attr-name"]`.  CEL variables can ecapsulate the `dev.attrs["example.com"]` part, yielding `xattrs["attrName"]`.

* Option 4: Split attr names on `/` and document both field and map lookups
Same as options 2 + 3, with the caveat that only attrs whose names are C identifiers can use fields.

* Option 5: One of above + fully qualifier attrs
Pick from options 2, 3, 4 and also allow `dev.attrs["example.com/attrName"]` or `dev.attrs["example.com/attr-name"]`

> it may help with simple cases (like `"dra.example.com/driverVersion" in dev.attrs`)

The alternative being `"dra.example.com" in dev.attrs && "driverVersion" in dev.attrs["dra.example.com"]` ?  Yikes.  Figuring out a sane dialect of CEL is tricky.   Can't we find a way to let non-existient attributes just always evaluate to false?

I wanted to try this, so I went to the CEl playground:

https://playcel.undistro.io/?content=H4sIAAAAAAAAA6tWSk7NUbJSykgs1khJLdNLLCkpKo6OUUqtSMwtyEnVS87PjVGK1VRQU1PALa2XlFikYGurEKMEZMQoxeQp6SilJJYkeuYVlJYADQfqtIrJU1AA6wazFBTQzICKgsTT8vOBfIRhMPGy1KLizPw8kJwl2Irc%2FJRUoOkgD9QCAP%2FhHsfIAAAA

The CEL spec says:

"""
To test for the presence of a field, the boolean-valued macro has(e.f) can be used.

If e evaluates to a map, then has(e.f) indicates whether the string f is a key in the map (note that f must syntactically be an identifier).
"""

There's no other affordance for e being a map and f NOT being an identifier.  Instead we have to do

`"example.com" in dev.attrs && has(dev.attrs["example.com"].foo) && dev.attrs["example.com"].foo == "bar"`  

Which is pretty god-awful.  On that I guess I support both being available.  https://playcel.undistro.io/?content=H4sIAAAAAAAAA22PwQrCMAyGXyXksJN0Zwt7AJ%2FBeKhrhMHajK4OQXx30zIZk92S%2F8%2F3J3ljzyNaJOSXC9PIppfQPkQIYYjgeTEu5zRD02zNdTdNeDMKQNcB4d0lQop4Qu%2Byu8TpmTVdSUsRoNK1AvjLWNWi1%2B12C%2FvpC6d5kFi880FGu3KFOnD3tB4YxLPeVv7%2FfAHHNuSJBwEAAA%3D%3D

The playground doesn't seem to support `bind`.

Reply:
> >    it may help with simple cases (like "dra.example.com/driverVersion" in dev.attrs)
> The alternative being "dra.example.com" in dev.attrs && "driverVersion" in dev.attrs["dra.example.com"] ? Yikes.

Yes, exactly that is why I brought it up.

> Figuring out a sane dialect of CEL is tricky. Can't we find a way to let non-existient attributes just always evaluate to false?

That is the "lookup with defaults" approach from the Google doc, which was rejected earlier. The problem with defaults for the leaf is that the type is unknown, so we cannot return a boolean false when the users expects it, because with untyped lookup we don't know what type the user expects. If we return a boolean when an int is needed, it's still going to be a runtime error.

What we can do is always return the empty map for the domain lookup. Then the "does it exist" check becomes `"driverVersion" in dev.attrs["dra.example.com"]`.

Give that `xattrs.attrName` looks natural, I am leaning towards option 2 with this "empty map for domain lookup" as default.

There is one small implementation detail to consider: should `"dra.example.com" in dev.attrs` return `true` or `false` when there is no attribute with that domain? `false` seems what someone would expect, except that `dev.attrs["dra.example.com"]` isn't wrong.

Reply:
> Give that xattrs.attrName looks natural, I am leaning towards option 2 with this "empty map for domain lookup" as default.

That could work. A few other ideas:

1) Allow an evaluation "option" (likely in the CEL struct) to provide simplified, default behavior where we validate any attribute references, and evaluate the *entire expression* to false if any looked up attribute does not exist. This means no guards are needed at all. If you want to do something complex with guards (`"x" in attrs ? attrs["x"] : attrs["y"]`), you would need to turn off the default-on option. This results in more compact expressions since no guards are needed, at the expense of "magic" that may be non-obvious coming from other parts of K8s.

2) We eliminate the implicit magic of 1) by making the user list attributes that they expect to exist, and marking the expression false if they don't exist. Sort of explicit assertions. It keeps the expressions a little simpler but is still a pain.

I think I like the default empty map best at this point, but figured I'd raise those ideas nonetheless. There may be some other things we can think of.

We might want to find some way to surface failed lookups, though, because otherwise a typo may be hard to detect. This would probably be in the claim status. One option would be that if we find an attribute not defined by *any* device in the cluster, it's probably a typo and we can surface that in the status message. Another (not mutually exclusive) option would be to include statistics on each attribute used in the expression (`attributes["example.com"].foo exists in 3 of 10,245 evaluated devices`). That could give folks a hint that they made an error (if it exists in 0 devices).

By the way, not to pull on yet another "decided" thread, but if we treated "driver" as special, and allowed it to optionally be specified up-front rather than in the CEL expression, we could get some advantages here both in early pruning of evaluations and perhaps in CEL expressions ("if you specify the driver up front, you don't have to qualify the names"?).

Reply:
Your two options sound like future extensions that might get added after we have more experience with a simpler approach. The same applies to the ideas how to surface problems: this is too experimental to include it in 1.31.

> That could give folks a hint that they made an error (if it exists in 0 devices).

We would need to be careful to not phrase it as an error, because it is legitimate to deploy a portable app in a cluster where only one of two alternatives exists.

> if we treated "driver" as special, and allowed it to optionally be specified up-front rather than in the CEL expression

I had that in one of my early drafts. We removed it as "too much magic". Let's not go back there.

Regarding field lookup vs. map lookup for the leaf: elsewhere today @klueska made a good point that the field syntax makes the field look too much like something that is standardized by Kubernetes. It's the same problem we discussed with plain maps in the Kubernetes API: there it looks like the keys are part of the schema, here it's the field names.

`dev.attrs["dra.example.com"]["memory"]` avoids that false impression by treating both as strings. So let me take back what I said earlier, I now prefer two-level map access over map+field.

Reply:
>> That could give folks a hint that they made an error (if it exists in 0 devices).
>
> We would need to be careful to not phrase it as an error, because it is legitimate to deploy a portable app in a cluster where only one of two alternatives exists.

Agree and agree - I think we would not surface general non-matches (too many), but zero matches is sus (as my kids would say).  Question:  would we surface zero-matches if there were other conditionals that did match?  e.g. `dev.attrs.goodName == "real" || dev.attr.typoNam == "oops"` - if that matches some devices, but we don't surface zero-match cases, we might never notice the typo, but it is valid to not match on one branch of a conditional.  

> if we treated "driver" as special, and allowed it to optionally be specified up-front rather than in the CEL expression

Are we putting `driverName` in the `attrs` or should it be `dev.driver` ?

> Regarding field lookup vs. map lookup for the leaf: elsewhere today @klueska made a good point that the field syntax makes the field look too much like something that is standardized by Kubernetes. It's the same problem we discussed with plain maps in the Kubernetes API: there it looks like the keys are part of the schema, here it's the field names.

I disagree that it is problematic in this context - it's not as if there's actually a struct with this schema (`dev.attrs[*].*`) anywhere in the system - it is purely synthetic.  Also, in hindsight, the restrictions on maps was almost certainly wrong, and I think we would have done it differently if we could go back.  That said, we can't actually PREVENT field lookups, as long as the name is a valid identifier.  I slightly prefer the field-style lookup, but @pohly, I'll let you own this one.  Perhaps a compormise is the require C identifiers for names, and as we progress thru alpha and even beta, we can decide, WITH REAL USER INPUT, which style makes people happier?

Reply:
> Question: would we surface zero-matches if there were other conditionals that did match?

I don't think the scheduler would notice. It's not getting a runtime error if the first term of the `||` is true. This looks like a case for a "CEL linter" which looks at the AST and with better knowledge about which attributes are valid reports problems.

As I suggested in the implementation PR, a `DeviceClass` should have a `spec` and a `status` like a `ResourceClaim`. Then a condition can be used to mark "bad" classes and claims by such a linter that is running in the cluster.

> Are we putting driverName in the attrs or should it be dev.driver ?

We have `device.driverName`, which exposes `slice.spec.driverName`, so that is not an attribute.

What @johnbelamaric alluded to is putting a `DriverName string` next to the `Expression string` in the CEL selector. That can do some pre-filtering without evaluating the CEL expression and then if in the CEL expression there is ambiguity, it could be resolved as "use attributes in that driver's domain".  But as I said, too much magic.

> I slightly prefer the field-style lookup, but @pohly, I'll let you own this one.

Let's go with field-style if you aren't concerned about this aspect and because `variable.field` looks better. `dev.attrs["dra.k8s.io"]["numa"]` also could raise the question why `dev.attrs["dra.k8s.io/numa"]` isn't supported when that is how it is listed in a slice.

Reply:
SGTM

Linters being async makes me a little unhappy, but I think we could revisit that if evidence suggests the need

Reply:
Duh!

I just rediscovered what @thockin mentioned earlier ("we cannot prevent field lookups")....

Reply:
> I'll be damned if I can find the spec Joe linked me to in the past) idea, you get something like:
> `bind(attrs, dev.attrs["example.com"], attrs.model == "foo" && attrs.version > "1.2.3")`

Found it: https://pkg.go.dev/github.com/google/cel-go/ext#Bindings

It has to be enabled explicitly, so I am going to do that for the CEL selector environment.

Reply:
Implemented in https://github.com/kubernetes/kubernetes/pull/125488.